README.txt
= ImageScience

* http://seattlerb.rubyforge.org/ImageScience.html
* http://rubyforge.org/projects/seattlerb

== DESCRIPTION:

ImageScience is a clean and happy Ruby library that generates
thumbnails -- and kicks the living crap out of RMagick. Oh, and it
doesn't leak memory like a sieve. :)

For more information including build steps, see http://seattlerb.rubyforge.org/

== FORK

This is a fork of seattlerb/image_science.
Converted to use extconf instead of ruby-inline.
Added many new features.
See History.txt for the full list of changes since the fork.

== FEATURES/PROBLEMS:

* Glorious graphics manipulation magi... errr, SCIENCE! in less than 200 LoC!
* Supports square and proportional thumbnails, as well as arbitrary resizes.
* Pretty much any graphics format you could want. No really.

== SYNOPSYS:

  ImageScience.with_image(file) do |img|
    img.cropped_thumbnail(100) do |thumb|
      thumb.save "#{file}_cropped.png"
    end

    img.thumbnail(100) do |thumb|
      thumb.save "#{file}_thumb.png"
    end
  end

== REQUIREMENTS:

* FreeImage
* ImageScience

== INSTALL:

* Download and install FreeImage. See notes at url above.
* sudo gem install -y image_science
* see http://seattlerb.rubyforge.org/ImageScience.html for more info.

== LICENSE:

(The MIT License)

Copyright (c) 2006-2009 Ryan Davis, Seattle.rb

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
== 1.1.7 / 2011-02-05

* backward compatibility fix for ruby 1.8.7
* fix for compilation under rvm on OS X

== 1.1.6 / 2011-01-10

* merged in dpi code from tcaddy (https://github.com/tcaddy/image_science)
** dpm_to_dpi, dpi_to_dpm, dpi_x, dpi_y, dpi, resize_with_dpm, dpm_x, dpm_y

== 1.1.5 / 2010-12-16

* fixes for ruby 1.9

== 1.1.4 / 2010-09-15

* added 'fit_within' command to bin/image_science.
* added options information to bin/image_science -h
* merged changes from scotje/image_science - fix for gif conversion to 8bit

== 1.1.3 / 2010-02-23

* fixed parameters to FreeImage_Rotate

== 1.1.2 / 2010-02-12

* added bin/image_science script.
* added buffer() method to get image data as a string.
* crop(), with_crop(), resize(), thumbnail(), cropped_thumbnail() can now all be called without providing a block (modifies the image in-place).
* resize() now accepts optional filter (see FreeImage::ImageFilters).
* image_type() can now be called as an instance method.
* added bmp, tiff, xpm files to test suite.
* added support for ImageScience.new (takes filename or data)
* added flip_horizontal, flip_vertical, rotate methods
* added support for load/save flags to new() and buffer().
* added set_pixel_color method
* added fit_within method

== 1.1.1 / 2010-02-06

* merged changes from scambra/master (depth, colorspace, image_type, colortype, file_type methods)
* reinstated lib/image_science.rb for methods implemented in ruby, c extension 
is now named image_science_ext
* added histogram, invert, adjust_contrast, adjust_brightness, adjust_gamma
* fixed up rdoc
* defines various FreeImage constants
* generate ext/image_science_ext.c from ext/image_science_ext.c.in

== 1.1.0 / 2010-02-05

* converted to use extconf (removed ruby-inline)
* converted tests to rspec
* added tests for different file types (gif, png, jpg)

== 1.0.0 / 2010-01-14

* Fork of seattlerb/image_science
* added get_pixel_color method 
* build using LDFLAGS from rbconfig.rb


spec/spec_helper.rb
begin
  require 'spec'
rescue LoadError
  require 'rubygems'
  gem 'rspec'
  require 'spec'
end

$:.unshift(File.dirname(__FILE__) + '/../lib')
require 'cp-deployment'


spec/image_science_spec.rb
require File.dirname(__FILE__) + '/../lib/image_science'

include ImageScience::ColorChannels
include ImageScience::ColorTypes
include ImageScience::ImageFilters
include ImageScience::ImageFormats

describe ImageScience do

  FILE_TYPES = %W{png jpg gif bmp tif xpm}

  before(:each) do
    @path = 'spec/fixtures'
    @h = @w = 50
  end

  after(:each) do
    FILE_TYPES.each do |ext|
      File.unlink tmp_image_path(ext) if File.exist? tmp_image_path(ext)
    end
  end

  describe "get_version" do
    it "should return the image science version" do
      ImageScience.get_version.should_not be_nil
    end
  end
  
  FILE_TYPES.each do |ext|

    describe "#{ext}" do

      describe "new" do
        it "should create a new ImageScience object by reading a file" do
          img = ImageScience.new(image_path(ext))
          img.should be_kind_of(ImageScience)
          [img.width, img.height].should == [@w, @h]
        end

        it "should create a new ImageScience object by reading image data" do
          data = File.read(image_path(ext))
          img = ImageScience.new(data)
          img.should be_kind_of(ImageScience)
          [img.width, img.height].should == [@w, @h]
        end
      end

      describe "with_image" do
        it "should raise an error when a file does not exist" do
          lambda {
            ImageScience.with_image(image_path(ext) + "nope") {}
          }.should raise_error
        end

        it "should fetch image dimensions" do
          ImageScience.with_image image_path(ext) do |img|
            img.should be_kind_of(ImageScience)
            img.height.should == @h
            img.width.should == @w
          end
        end
      end

      describe "with_image_from_memory" do
        it "should raise an error when an empty string is given" do
          lambda {
            ImageScience.with_image_from_memory("") {}
          }.should raise_error
        end
      end

      describe "with_image_from_memory" do
        it "should fetch image dimensions" do
          data = File.new(image_path(ext)).binmode.read
          ImageScience.with_image_from_memory data do |img|
            img.should be_kind_of(ImageScience)
            img.height.should == @h
            img.width.should == @w
          end
        end
      end

      describe "save" do
        it "should save a new copy of an image" do
          ImageScience.with_image image_path(ext) do |img|
            img.save(tmp_image_path(ext)).should be_true
          end
          File.exists?(tmp_image_path(ext)).should be_true
          
          ImageScience.with_image tmp_image_path(ext) do |img|
            img.should be_kind_of(ImageScience)
            img.height.should == @h
            img.width.should == @w
          end
        end
      end

      describe "resize" do
    
        it "should resize an image" do
          ImageScience.with_image image_path(ext) do |img|
            img.resize(25, 25) do |thumb|
              thumb.save(tmp_image_path(ext)).should be_true
            end
          end

          File.exists?(tmp_image_path(ext)).should be_true

          ImageScience.with_image tmp_image_path(ext) do |img|
            img.should be_kind_of(ImageScience)
            img.height.should == 25
            img.width.should == 25
          end
        end

        it "should resize an image given floating point dimensions" do
          ImageScience.with_image image_path(ext) do |img|
            img.resize(25.2, 25.7) do |thumb|
              thumb.save(tmp_image_path(ext)).should be_true
            end
          end

          File.exists?(tmp_image_path(ext)).should be_true
        
          ImageScience.with_image tmp_image_path(ext) do |img|
            img.should be_kind_of(ImageScience)
            img.height.should == 25
            img.width.should == 25
          end
        end
      
        # do not accept negative or zero values for width/height
        it "should raise an error if given invalid width or height" do
          [ [0, 25], [25, 0], [-25, 25], [25, -25] ].each do |width, height|
            lambda {
              ImageScience.with_image image_path(ext) do |img|
                img.resize(width, height) do |thumb|
                  thumb.save(tmp_image_path(ext))
                end
              end
            }.should raise_error
            
            File.exists?(tmp_image_path(ext)).should be_false
          end
        end

        it "should resize the image in-place if no block given" do
          ImageScience.with_image image_path(ext) do |img|
            img.resize(20, 20).should be_true
            img.width.should == 20
            img.height.should == 20
          end
        end

        it "should resize the image with the given filter" do
          ImageScience.with_image image_path(ext) do |img|
            img.resize(20, 20, FILTER_BILINEAR).should be_true
            img.width.should == 20
            img.height.should == 20
          end
        end
        
      end

      describe "buffer" do
        it "should return image data" do
          ImageScience.with_image image_path(ext) do |img|
            expected = File.size(image_path(ext))
            tolerance = expected * 0.05

            data = img.buffer
            data.should_not be_nil
            #data.length.should be_close(expected, tolerance)
          end
        end

        it "should yield image data" do
          ImageScience.with_image image_path(ext) do |img|
            expected = File.size(image_path(ext))
            tolerance = expected * 0.05

            img.buffer do |data|
              data.should_not be_nil
              #data.length.should be_close(expected, tolerance)
            end
          end
        end

        it "should return image data in the given format" do
          ImageScience.with_image image_path(ext) do |img|
            [FIF_BMP, FIF_GIF, FIF_JPEG, FIF_PNG].each do |target_format|
              data = img.buffer(target_format)
              data.should_not be_nil
            end
          end
        end

        # TODO: convert to use constants
        it "should accept save flags" do
          ImageScience.with_image image_path(ext) do |img|
            jpg_large = img.buffer(FIF_JPEG, 0x80)  # jpeg quality superb
            jpg_small = img.buffer(FIF_JPEG, 0x08)  # jpeg quality bad
            jpg_small.length.should < jpg_large.length
          end
        end

      end

      describe "fit_within" do

        it "should resize image to fit within given dimensions (yield)" do
          # 50 x 50 image -> shrink to 20, 50 -> 20, 20
          ImageScience.with_image image_path(ext) do |img|
            img.fit_within(20, 50) do |thumb|
              thumb.width.should == 20
              thumb.height.should == 20
            end
          end
        end

        it "should resize image to fit within given dimensions (inline)" do
          # 50 x 50 image -> shrink to 20, 50 -> 20, 20
          ImageScience.with_image image_path(ext) do |img|
            img.fit_within(20, 50)
            img.width.should == 20
            img.height.should == 20
          end

          # 100 x 50 image -> shrink to 50, 100 -> 50, 25
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.fit_within(50, 100)
            img.width.should == 50
            img.height.should == 25
          end

          # 100 x 50 image -> shrink to 150, 25 -> 50, 25
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.fit_within(150, 25)
            img.width.should == 50
            img.height.should == 25
          end
        end

      end

      describe "get_pixel_color" do
        expected = {
          :jpg => [[62, 134, 122, 0],  [0, 11, 6, 0]],
          :png => [[62, 134, 121, 0],  [1, 2, 2, 0]],
          :gif => [[59, 135, 119, 0],  [0, 2, 0, 0]],
          :bmp => [[62, 134, 121, 0],  [1, 2, 2, 0]],
          :tif => [[62, 134, 121, 0],  [1, 2, 2, 0]],
          :xpm => [[255, 255, 255, 0], [0, 0, 0, 0]]
        }

        it "should get pixel color" do
          ImageScience.with_image image_path(ext) do |img|
            rgb = img.get_pixel_color(10,7)
            rgb.should_not be_nil
            rgb.should == expected[ext.to_sym][0]
          
            rgb = img.get_pixel_color(24,0)
            rgb.should_not be_nil
            rgb.should == expected[ext.to_sym][1]
          end
        end

        it "should get pixel color with []" do
          ImageScience.with_image image_path(ext) do |img|
            rgb = img[10,7]
            rgb.should_not be_nil
            rgb.should == expected[ext.to_sym][0]
          
            rgb = img[24,0]
            rgb.should_not be_nil
            rgb.should == expected[ext.to_sym][1]
          end
        end
      end

      describe "set_pixel_color" do
        it "should set pixel color" do
          ImageScience.with_image image_path(ext) do |img|
            if img.colortype == FIC_PALETTE
              img.set_pixel_color(10, 7, 3).should be_true
              img[25, 25] = 4
            else
              img.set_pixel_color(10,  7, [100, 20, 50]).should be_true
              img.set_pixel_color(25, 25, 100, 20, 50).should be_true
              img.set_pixel_color(25, 25, [100, 20, 50, 0]).should be_true
              img.set_pixel_color(25, 25, 100, 20, 50, 0).should be_true
              img[25, 25] = [100, 20, 50, 0];
            end
          end
        end
      end
      
      describe "thumbnail" do
        # Note: pix2 is 100x50
        it "should create a proportional thumbnail" do
          thumbnail_created = false
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.thumbnail(30) do |thumb|
              thumb.should_not be_nil
              thumb.width.should  == 30
              thumb.height.should == thumb.width / 2 # half of width
              thumbnail_created = true
            end
          end
          thumbnail_created.should be_true
        end

        it "should create a proportional thumbnail in-place if no block given" do
          thumbnail_created = false
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.thumbnail(30)
            img.width.should  == 30
            img.height.should == img.width / 2 # half of width
            thumbnail_created = true
          end
          thumbnail_created.should be_true
        end
      end
      
      describe "cropped_thumbnail" do
        # Note: pix2 is 100x50
        it "should create a square thumbnail" do
          thumbnail_created = false
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.cropped_thumbnail(30) do |thumb|
              thumb.should_not be_nil
              thumb.width.should == 30
              thumb.height.should == 30   # same as width
              thumbnail_created = true
            end
          end
          thumbnail_created.should be_true
        end

        it "should create a square thumbnail in-place if no block given" do
          thumbnail_created = false
          ImageScience.with_image image_path(ext, "pix2") do |img|
            img.cropped_thumbnail(30)
            img.width.should == 30
            img.height.should == 30   # same as width
            thumbnail_created = true
          end
          thumbnail_created.should be_true
        end
      end

      describe "crop" do
        it "should crop the image in-place if no block given" do
          ImageScience.with_image image_path(ext) do |img|
            img.crop(0, 0, 25, 20).should be_true
            img.width.should == 25
            img.height.should == 20
          end
        end
      end

      # image_type calls ImageScience.file_type, converts to string.
      # allow calling as a class or instance method
      describe "image_type" do
        expected = {
          'gif' => 'GIF',
          'jpg' => 'JPEG',
          'png' => 'PNG',
          'bmp' => 'BMP',
          'tif' => 'TIFF',
          'xpm' => 'XPM'
        }
        it "should return the image type (class method)" do
          ImageScience.image_type(image_path(ext)).should == expected[ext]
        end

        it "should return the image type (instance method)" do
          ImageScience.with_image image_path(ext) do |img|
            img.image_type.should == expected[ext]
          end
        end
      end

      # colorspace calls img.colortype & img.depth, converts to string
      describe "colorspace" do
        it "should return the color space" do
          expected = {
            'gif' => 'Indexed',
            'jpg' => 'RGB',
            'png' => 'RGB',
            'bmp' => 'RGB',
            'tif' => 'RGB',
            'xpm' => 'Indexed'
          }
          ImageScience.with_image image_path(ext) do |img|
            img.colorspace.should == expected[ext]
          end
        end
      end

      describe "depth" do
        it "should return the BPP of the image" do
          expected = {
            'gif' => 8,
            'jpg' => 24,
            'png' => 24,
            'bmp' => 24,
            'tif' => 24,
            'xpm' => 8
          }
          ImageScience.with_image image_path(ext) do |img|
            img.depth.should == expected[ext]
          end
        end
      end

      describe "adjust_gamma" do
        it "should perform gamma correction" do
          ImageScience.with_image image_path(ext) do |img|
            # darken image
            img.adjust_gamma(0.5).should be_true
          end
        end
      end

      describe "adjust_brightness" do
        it "should adjust brightness" do
          ImageScience.with_image image_path(ext) do |img|
            # 50% brighter
            img.adjust_brightness(50).should be_true
          end
        end
      end

      describe "adjust_contrast" do
        it "should adjust contrast" do
          ImageScience.with_image image_path(ext) do |img|
            # 50% less contrast
            img.adjust_contrast(-50).should be_true
          end
        end
      end

      describe "invert" do
        it "should invert pixel data" do
          ImageScience.with_image image_path(ext) do |img|
            # 50% less contrast
            img.invert.should be_true
          end
        end
      end

      describe "histogram" do
        it "should compute the image histogram" do
          ImageScience.with_image image_path(ext) do |img|
            h = img.histogram
            h.should be_kind_of(Array)
            h.length.should == 256
          end
        end

        it "should compute the image histogram for a given channel" do
          ImageScience.with_image image_path(ext) do |img|
            [FICC_RED, FICC_GREEN, FICC_BLUE].each do |channel|
              h = img.histogram(channel)
              h.should be_kind_of(Array)
              h.length.should == 256
            end
          end
        end
      end

      describe "rotate" do

        # test Rotate
        it "should rotate the image 90 degrees" do
          ImageScience.with_image image_path(ext, "pix2") do |img|
            w, h = img.width, img.height
            img.rotate(90)
            # width & height reversed
            img.width.should == h
            img.height.should == w
          end
        end

        # test RotateEx
        it "should rotate the image 120 degrees about an origin" do
          ImageScience.with_image image_path(ext, "pix2") do |img|
            w, h = img.width, img.height
            img.rotate(120, 0, 0, 20, 20)
            # width & height unchanged
            img.width.should == w
            img.height.should == h
          end          
        end

      end

      describe "flip_horizontal" do
        it "should flip the image horizontally" do
          ImageScience.with_image image_path(ext) do |img|          
            img.flip_horizontal.should be_true
          end
        end
      end

      describe "flip_vertical" do
        it "should flip the image vertically" do
          ImageScience.with_image image_path(ext) do |img|          
            img.flip_vertical.should be_true
          end
        end
      end

      # dpi methods (from tcaddy)

      describe "dpm_x" do        
        it "should return the dpm on the x axis" do
          ImageScience.with_image image_path(ext) do |img|
            img.dpi_x.should == 72.009
          end
        end
      end

      describe "dpm_y" do
        it "should return the dpm on the y axis" do
          ImageScience.with_image image_path(ext) do |img|
            img.dpi_y.should == 72.009
          end
        end
      end

      describe "resize_with_dpm" do
        it "should resize the image to the specified dpm" do
          ImageScience.with_image image_path(ext) do |img|
            img.resize_with_dpm(30, 30, 100, 100).should be_true
            img.dpm_x.should == 100
            img.dpm_y.should == 100
          end
        end
      end

    end
  end

  private

  def image_path(extension, basename = "pix")
    raise "extension required" unless extension
    File.join(@path, "#{basename}.#{extension}")
  end

  def tmp_image_path(extension, basename = "pix")
    raise "extension required" unless extension
    File.join(@path, "#{basename}-tmp.#{extension}")
  end

end



bench.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'benchmark'
require 'image_science'

max = (ARGV.shift || 100).to_i
ext = ARGV.shift || "png"

file = "blah_big.#{ext}"

if RUBY_PLATFORM =~ /darwin/ then
  # how fucking cool is this???
  puts "taking screenshot for thumbnailing benchmarks"
  system "screencapture -SC #{file}"
else
  abort "You need to plonk down #{file} or buy a mac"
end unless test ?f, "#{file}"

ImageScience.with_image(file.sub(/#{ext}$/, 'png')) do |img|
  img.save(file)
end if ext != "png"

puts "# of iterations = #{max}"
Benchmark::bm(20) do |x|
  x.report("null_time") {
    for i in 0..max do
      # do nothing
    end
  }

  x.report("cropped") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.cropped_thumbnail(100) do |thumb|
          thumb.save("blah_cropped.#{ext}")
        end
      end
    end
  }

  x.report("proportional") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.thumbnail(100) do |thumb|
          thumb.save("blah_thumb.#{ext}")
        end
      end
    end
  }

  x.report("resize") {
    for i in 0..max do
      ImageScience.with_image(file) do |img|
        img.resize(200, 200) do |resize|
          resize.save("blah_resize.#{ext}")
        end
      end
    end
  }
end

# File.unlink(*Dir["blah*#{ext}"])


ext/extconf.rb
require 'mkmf'

# expand comments in image_science_ext.c.in and generate image_science_ext.c.
# creates constant definitions (rb_define_const)
def expand_constants

  File.open("conftest.c", "w") { |f| f.puts "#include <FreeImage.h>" }
  cpp = cpp_command('')
  system "#{cpp} > confout"

  constants = {}
  headers = []

  config = File.readlines("confout")
  config.each do |include|
    next unless include.match(/"(.*?FreeImage.h)"/)
    filename = include.split('"')[1]
    headers << filename
  end
  
  headers.uniq!

  # add typedef constants
  config.each do |define|
    next unless define.match(/^\s*(\w+)\s*=\s*\d/)  # typedef
    name = $1
    next unless name.match(/^(FIT|FICC|FIC|FIF|FILTER)_/)
    constants[$1] ||= []
    constants[$1] << [name]
  end

  if RUBY_VERSION >= "1.9"
    raw_headers = headers.collect { |i| File.read(i, :encoding => "ISO-8859-1") }.join
  else
    raw_headers = headers.collect { |i| File.read(i) }.join
  end
  
  # add #defined constants (load/save flags)
  constants['FLAG'] ||= []
  flag_defines = raw_headers.scan(/(flag constants\s*---.*?)^\/\//m)
  flag_defines.each do |flag_data|
    flag_data[0].split(/\n/).each do |define|
      next unless define.match(/^\#define\s*(\w+)\s(.*?(\/\/.+)$)?/)  # #define
      constants['FLAG'] << [$1, $3]
    end
  end

  File.unlink("confout")

  constants.keys.each { |i| constants[i].uniq! }

  File.open("image_science_ext.c", "w") do |newf|
    File.foreach("image_science_ext.c.in") do |l|
      if l.match(/\/\* expand FreeImage constants\s+(\w+)\s+(\w+)\s*\*\//)
        klass_name = $1
        const_type = $2
        const_list = constants[const_type]
        unless const_list
          puts "warning: no constants found matching #{const_type}"
          next
        end
        const_list.each do |c, comment|
          if comment
            comment.sub!('//', '')
            comment.sub!(/[\r\n]+$/, '')
            comment.gsub!(/:/, '-')  # colons break rdoc..
            comment.strip!
            newf.puts %Q{  /* #{comment} */}
          end
          newf.puts %Q{  rb_define_const(#{klass_name}, "#{c}", INT2FIX(#{c}));}
        end
      else
        newf.puts l
      end
    end
  end

end

dir_config('freeimage', %W{/opt/local /usr/local /usr})

ok = have_header('FreeImage.h') &&
  have_library('stdc++') && # sometimes required on OSX
  have_library('freeimage', 'FreeImage_Load')

if(ok)
  have_func('FreeImage_Rotate')
  have_func('FreeImage_RotateClassic')
  expand_constants
  create_makefile("image_science_ext")
end


ext/image_science_ext.c
/*
 * Provides a clean and simple API to generate thumbnails using
 * FreeImage as the underlying mechanism.
 *
 * For more information or if you have build issues with FreeImage, see
 * http://seattlerb.rubyforge.org/ImageScience.html
 *
 */

#include "ruby.h"
#include "FreeImage.h"

#define GET_BITMAP(name) Data_Get_Struct(self, FIBITMAP, (name)); if (!(name)) rb_raise(rb_eTypeError, "Bitmap has already been freed");

VALUE isc;  /* ImageScience class */

static VALUE unload(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  FreeImage_Unload(bitmap);
  DATA_PTR(self) = NULL;
  return Qnil;
}

static VALUE wrap_and_yield(FIBITMAP *image, VALUE self, FREE_IMAGE_FORMAT fif) {
  unsigned int self_is_class = rb_type(self) == T_CLASS;
  VALUE klass = self_is_class ? self         : CLASS_OF(self);
  VALUE type  = self_is_class ? INT2FIX(fif) : rb_iv_get(self, "@file_type");
  VALUE obj = Data_Wrap_Struct(klass, NULL, NULL, image);
  rb_iv_set(obj, "@file_type", type);
  return rb_ensure(rb_yield, obj, unload, obj);
}

static void copy_icc_profile(VALUE self, FIBITMAP *from, FIBITMAP *to) {
  FREE_IMAGE_FORMAT fif = FIX2INT(rb_iv_get(self, "@file_type"));
  if (fif != FIF_PNG && FreeImage_FIFSupportsICCProfiles(fif)) {
    FIICCPROFILE *profile = FreeImage_GetICCProfile(from);
    if (profile && profile->data) {
      FreeImage_CreateICCProfile(to, profile->data, profile->size);
    }
  }
}

static void FreeImageErrorHandler(FREE_IMAGE_FORMAT fif, const char *message) {
  rb_raise(rb_eRuntimeError,
	   "FreeImage exception for type %s: %s",
	   (fif == FIF_UNKNOWN) ? "???" : FreeImage_GetFormatFromFIF(fif),
	   message);
}

/****** Class methods ******/

/*
 * The top-level image loader opens +path+ and then yields the image.
 */
static VALUE with_image(VALUE klass, VALUE filename) {
  FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;
  int flags;
  char *input = RSTRING(filename)->ptr;

  fif = FreeImage_GetFileType(input, 0);
  if (fif == FIF_UNKNOWN) fif = FreeImage_GetFIFFromFilename(input);
  if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsReading(fif)) {
    FIBITMAP *bitmap;
    VALUE result = Qnil;
    flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;
    bitmap = FreeImage_Load(fif, input, flags);
    if (bitmap) {
      FITAG *tagValue = NULL;
      FreeImage_GetMetadata(FIMD_EXIF_MAIN, bitmap, "Orientation", &tagValue); 
      switch (tagValue == NULL ? 0 : *((short *) FreeImage_GetTagValue(tagValue))) {
      case 6:
	bitmap = FreeImage_RotateClassic(bitmap, 270);
	break;
      case 3:
	bitmap = FreeImage_RotateClassic(bitmap, 180);
	break;
      case 8:
	bitmap = FreeImage_RotateClassic(bitmap, 90);
	break;
      default:
	break;
      }

      result = wrap_and_yield(bitmap, klass, fif);
    }
    return result;
  }
  rb_raise(rb_eTypeError, "Unknown file format");
}

/*
 * The top-level image loader, opens an image from the string +data+
 * and then yields the image.
 */
static VALUE with_image_from_memory(VALUE klass, VALUE image_data) {
  FREE_IMAGE_FORMAT fif = FIF_UNKNOWN;

  Check_Type(image_data, T_STRING);
  BYTE *image_data_ptr    = (BYTE*)RSTRING_PTR(image_data);
  DWORD image_data_length = RSTRING_LEN(image_data);
  FIMEMORY *stream = FreeImage_OpenMemory(image_data_ptr, image_data_length);

  if (NULL == stream) {
    rb_raise(rb_eTypeError, "Unable to open image_data");
  }

  fif = FreeImage_GetFileTypeFromMemory(stream, 0);
  if ((fif == FIF_UNKNOWN) || !FreeImage_FIFSupportsReading(fif)) {
    rb_raise(rb_eTypeError, "Unknown file format");
  }

  FIBITMAP *bitmap = NULL;
  VALUE result = Qnil;
  int flags = fif == FIF_JPEG ? JPEG_ACCURATE : 0;
  bitmap = FreeImage_LoadFromMemory(fif, stream, flags);
  FreeImage_CloseMemory(stream);
  if (bitmap) {
    result = wrap_and_yield(bitmap, klass, fif);
  }
  return result;
}

/*
 * get FreeImage library version
 */
static VALUE get_version(VALUE self) {
  const char *version = FreeImage_GetVersion();
  return rb_str_new2(version);
}

static VALUE file_type(VALUE self, VALUE filename) {
  char * input = RSTRING(filename)->ptr;
  FREE_IMAGE_FORMAT fif = FIF_UNKNOWN; 

  fif = FreeImage_GetFileType(input, 0); 
  if (fif == FIF_UNKNOWN) fif = FreeImage_GetFIFFromFilename(input); 
  return (fif == FIF_UNKNOWN) ? Qnil : INT2FIX(fif);
}

/*********** Instance methods ***********/

/*
 * Crops an image to +left+, +top+, +right+, and +bottom+ and then
 * yields the new image.
 */
static VALUE with_crop(VALUE self, VALUE lv, VALUE tv, VALUE rv, VALUE bv) {
  FIBITMAP *copy, *bitmap;
  VALUE result = Qnil;
  GET_BITMAP(bitmap);

  copy = FreeImage_Copy(bitmap, NUM2INT(lv), NUM2INT(tv),
			NUM2INT(rv), NUM2INT(bv));
  if (copy) {
    copy_icc_profile(self, bitmap, copy);
    result = wrap_and_yield(copy, self, 0);
  }
  return result;
}

/*
 * Returns the height of the image, in pixels.
 */
static VALUE height(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);
  int height = FreeImage_GetHeight(bitmap);
  return INT2FIX(height);
}

/*
 * Returns the width of the image, in pixels.
 */
static VALUE width(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);
  int width = FreeImage_GetWidth(bitmap);
  return INT2FIX(width);
}

/*
 * Returns an array representing the color of the given pixel (blue,green,red)
 */
static VALUE get_pixel_color(VALUE self, VALUE xval, VALUE yval) {
  FIBITMAP *bitmap;
  RGBQUAD rgb;
  RGBQUAD *pal;
  BYTE rgb_index;
  GET_BITMAP(bitmap);
  FREE_IMAGE_COLOR_TYPE ctype;
  VALUE out_ary = rb_ary_new2(3);
  int x = NUM2INT(xval);
  int y = NUM2INT(yval);
  int success = 0;

  ctype = FreeImage_GetColorType(bitmap);

  if(ctype == FIC_PALETTE) {
    if(FreeImage_GetPixelIndex(bitmap, x, y, &rgb_index)) {
      pal = FreeImage_GetPalette(bitmap);
      if(pal) {
	rgb = pal[rgb_index];
	success = 1;
      }
    }
  } else {
    success = FreeImage_GetPixelColor(bitmap, x, y, &rgb);
  }

  if(success) {
    rb_ary_store(out_ary, 0, INT2FIX(rgb.rgbRed));
    rb_ary_store(out_ary, 1, INT2FIX(rgb.rgbGreen));
    rb_ary_store(out_ary, 2, INT2FIX(rgb.rgbBlue));
  }

  return out_ary;
}

/***
 * Resizes the image to +width+ and +height+ using a cubic-bspline
 * filter and yields the new image.
 */
static VALUE resize(VALUE self, VALUE width, VALUE height) {
  FIBITMAP *bitmap, *image;
  int w = NUM2INT(width);
  int h = NUM2INT(height);

  if (w <= 0) rb_raise(rb_eArgError, "Width <= 0");
  if (h <= 0) rb_raise(rb_eArgError, "Height <= 0");
  GET_BITMAP(bitmap);
  image = FreeImage_Rescale(bitmap, w, h, FILTER_CATMULLROM);
  if (image) {
    copy_icc_profile(self, bitmap, image);
    return wrap_and_yield(image, self, 0);
  }
  return Qnil;
}

/*
 * Saves the image out to +path+. Changing the file extension will
 * convert the file type to the appropriate format.
 */
static VALUE save(VALUE self, VALUE filename) {
  int flags;
  char * output = RSTRING(filename)->ptr;
  FIBITMAP *bitmap;
  FREE_IMAGE_FORMAT fif = FreeImage_GetFIFFromFilename(output);
  if (fif == FIF_UNKNOWN) fif = FIX2INT(rb_iv_get(self, "@file_type"));
  if ((fif != FIF_UNKNOWN) && FreeImage_FIFSupportsWriting(fif)) {
    GET_BITMAP(bitmap);
    flags = fif == FIF_JPEG ? JPEG_QUALITYSUPERB : 0;
    BOOL result = 0, unload = 0;
    
    if (fif == FIF_PNG) FreeImage_DestroyICCProfile(bitmap);
    if (fif == FIF_JPEG && FreeImage_GetBPP(bitmap) != 24) {
      bitmap = FreeImage_ConvertTo24Bits(bitmap);
      unload = 1; // sue me
    }
    result = FreeImage_Save(fif, bitmap, output, flags);
    
    if (unload) FreeImage_Unload(bitmap);

    return result ? Qtrue : Qfalse;
  }
  rb_raise(rb_eTypeError, "Unknown file format");
}

static VALUE colortype(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  return INT2FIX(FreeImage_GetColorType(bitmap));
}

static VALUE depth(VALUE self) {
  FIBITMAP *bitmap;
  GET_BITMAP(bitmap);

  return INT2FIX(FreeImage_GetBPP(bitmap));
}

/* -- initialiser ---- */

void Init_image_science_ext(void)
{
  isc = rb_define_class("ImageScience", rb_cObject);
  rb_define_singleton_method(isc, "with_image", with_image, 1);
  rb_define_singleton_method(isc, "with_image_from_memory",
			     with_image_from_memory, 1);
  rb_define_singleton_method(isc, "get_version", get_version, 0);
  rb_define_singleton_method(isc, "file_type", file_type, 1);

  //rb_define_method(isc, "initialize", t_init, 0);
  rb_define_method(isc, "width", width, 0);
  rb_define_method(isc, "height", height, 0);
  rb_define_method(isc, "resize", resize, 2);
  rb_define_method(isc, "save", save, 1);
  rb_define_method(isc, "with_crop", with_crop, 4);
  rb_define_method(isc, "get_pixel_color", get_pixel_color, 2);
  rb_define_method(isc, "colortype", colortype, 0);
  rb_define_method(isc, "depth", depth, 0);

  FreeImage_SetOutputMessage(FreeImageErrorHandler);
}
